pretrain/dataset.py

```diff
@@ -275,28 +304,33 @@ def fetch_data_to_rows(self, num_pages):
                     )
                     raise
 
-
         return rows
 
     def get_random_pages(self, num_pages):
         """
-        Randomly sample one page.
-        A page is a row number of a given split of a given dataset dump.
+        Randomly sample num_pages.
+        A page is a row number of a given split of a given dataset dump offset by num_rows_per_page.
         """
         pages = []
 
         for _ in range(num_pages):
 
-            # Choose a random config
+            # Choose a random config.
             config_name = random.choice(list(self.configs_data.keys()))
 
-            # Choose a random page (row)
-            page = random.randint(0,
-                                  self.configs_data[config_name]['num_rows'] - 1 - self.num_rows_per_page)
+            # Choose a random page start.
```

Comment:
Not a big issue, but this reduces the effective size of the dataset. I think it divides it by `self.num_rows_per_page` which is 100.
we will never sample batches that span two adjacent chunks withing the same page. I think we divide the difficulty of overfitting the dataset by 100.
What I am saying is true when we evaluate on only one page, but becomes less and less true until you reach the extreme case where we evaluate on the whole dataset at every round.
A trick to solve this would be to pick up a random offset at each round, and then apply:

```
            selected_page_start = offset + (
                random.randint(0, data_page_count - 1) * self.num_rows_per_page
            )
```

Reply:
We currently do sample batches that span two adjacent chunks.

By adding this logic here we no longer do so.

It does not reduce the effective size by num rows per page because we read that number of pages. So we would get pages 0-99, 100-199 etc if we start at 0 or 100 right?

Reply:
Ah I think I see your point. You want it to sometimes get data from 50-149. Sure we could do that.

Reply:
Added code to do so. We pass in the initial offset so that we ensure across the overall dataset load we respect the same offset to avoid overlap, while still allowing each load to have a new offset.

Reply:
Yeah, exactly, the 50-149 idea ⚡